docs/design/features/OsrDetailsAndDebugging.md

```diff
@@ -0,0 +1,382 @@
+* The normal case is that a method is first quickly jitted (at Tier0)
+and then rejitted (at Tier1) (on a background thread) once the
+method has been called frequently.
+* If a method has been prejitted then the prejitted code serves
+the place of the Tier0 code.
+* Some methods bypass tiering and are immediately optimized
+the first time they are jitted:
+  * Methods marked with the `AggressiveOptimization` attribute
+  * Dynamic methods and methods from collectible assemblies
+  * By default in .NET 3.0, 5.0, 6.0: methods with loops.
+This policy is controlled by `DOTNET_COMPlus_TC_QuickJitForLoops`;
+ see next section.
+
+### Dynamic and Full PGO
+
+Dynamic PGO was introduced in the .NET 6.0 release. It also relies
+on tiered compilation.
+
+In Dynamic PGO, any method jitted at Tier0 has additional
+instrumentation code added to the Tier0 version to collect profile
+data.  This data is made available to JIT when the method is
+recompiled at Tier1.
+
+Full PGO is an offshoot of Dynamic PGO where we force all methods
+to pass through Tier0.
+
+### Quick Jit For Loops (aka `QJFL`)
+
+Tiered compilation was introduced in the .NET Core 3.0 release. 
+
+Initially, all methods that got jitted went through Tier0. But
+during the development cycle we got a fair amount of feedback that
+this had adverse performance impacts as performance-sensitive user
+code could be trapped in the Tier0 version.
+
+In response to this we changed the default behavior: methods with
+loops would be initially optimized and not participate in tiered jitting.
+We often use the shorthand `QJFL=0` to describe this behavior.
+
+While `QJFL=0` addressed the immediate problems users had with tiered
+compilation, it had some downsides: 
+
+* Not all methods with loops are performance sensitive, so in many
+cases we saw increased startup JIT time without any steady state
+benefit.  In some applications where startup JIT time is significant
+the impact was on the order or 20%.
+
+* In fact, early optimization of methods may lead to lower
+steady-state performance, because (a) class initializers may not yet
+have run, so early optimized methods must run class initializer checks
+and also cannot examine readonly static fields; (b) Dynamic PGO can
+only operate on methods that pass through Tier0.
+
+One can change the default behavior by setting QJFL to 1; some
+startup-sensitive applications (eg `Powershell`) do this.
+
+### Rationale for Enabling OSR
+
+OSR allows the runtime to jit most methods at Tier0, and to transition
+from Tier0 code when and if needed, and so allows us to change the
+default to `QJFL=1`.
+
+Enabling OSR will lead to:
+* better startup perf in general (upwards of 20% in some cases)
+* improved steady state perf
+* improved perf from Dynamic PGO (closing the gap between Dynamic
+ PGO and Full PGO)
+
+## How OSR Works
+
+### Enabling OSR
+
+OSR is enabled by setting
+`DOTNET_TC_QuickJitForLoops=1` (aka `QJFL=1`)
+`DOTNET_TC_OnStackReplacement=1` (aka `OSR=1`)
```

Comment:
Good question. I have thought about this some but don't have a great answer. Since we plan to have OSR=1 by default it probably doesn't matter, QJFL will gate things like it does now. 

QJFL shows through in project file settings and runtime json. I hope we won't need OSR to show up there as well.